File: src/main/java/nickleby/algorithms/domain/Elevators.java
package nickleby.algorithms.domain;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.IntStream;

public class Elevators {

    private static final Integer[] weights = {40, 40, 100, 80, 20};
    private static final Integer[] floors = {3, 3, 2, 2, 3};

    private static class Elevator {
        final Integer capacity;
        final Integer maxWeight;
        Integer currentWeight = 0;
        Integer currentCapacity = 0;
        Set<Integer> floors = new HashSet<>();
        Integer count = 0;

        Elevator(Integer capacity, Integer maxWeight) {
            this.capacity = capacity;
            this.maxWeight = maxWeight;
        }

        boolean isFull() { return capacity < currentCapacity + 1; }

        private boolean overWeight(Integer weight) {
            return maxWeight < currentWeight + weight;
        }

        void enter(Integer weight, Integer floor) {
            if(isFull() || overWeight(weight)) { dropOffOccupants(); }

            addOccupant(weight, floor);
        }

        void addOccupant(Integer weight, Integer floor) {
            floors.add(floor);
            currentCapacity++;
            currentWeight += weight;
        }

        void dropOffOccupants() {
            count += floors.size() + 1;
            floors = new HashSet<>();
            currentWeight = 0;
            currentCapacity = 0;
        }

        void finish() { this.dropOffOccupants(); }
    }

    public static void main(String[] args) {
        Elevator elevator = new Elevator(5, 200);

        IntStream.range(0, weights.length).boxed().forEach(i -> elevator.enter(weights[i], floors[i]));

        elevator.finish();

        System.out.println(elevator.count);
    }
}

File: src/main/java/nickleby/algorithms/kotlin/domain/Elevators.kt
package nickleby.algorithms.kotlin.domain

import java.util.ArrayList
import java.util.HashSet
import java.util.stream.IntStream

object Elevators {

  private val weights = arrayOf(40, 40, 100, 80, 20)
  private val floors = arrayOf(3, 3, 2, 2, 3)

  private class Elevator(var capacity: Int, var maxWeight: Int) {
    var weights: MutableList<Int> = ArrayList()
    var tempWeight: Int = 0
    var tempCapacity: Int = 0
    var floors: MutableSet<Int> = HashSet()
    var count: Int = 0

    fun leftLessThanRight(left: Int?, right: Int?, modifier: Int?) : Boolean {
      return left?.let { it < right!! + modifier!! }!!
    }

    fun enter(weight: Int?, floor: Int?): Elevator {
      if (leftLessThanRight(capacity, tempCapacity, 1) ||
          leftLessThanRight(maxWeight, tempWeight, weight))
      { finish() }

      weights.add(weight!!)
      floors.add(floor!!)
      tempCapacity++
      tempWeight += weight

      return this
    }

    fun finish(): Elevator {
      count += floors.size + 1
      weights = ArrayList()
      floors = HashSet()
      tempWeight = 0
      tempCapacity = 0

      return this
    }
  }

  @JvmStatic
  fun main(args: Array<String>) {
    val elevator = Elevator(5, 200)

    for(i in 0 until weights.size) { elevator.enter(weights[i], floors[i]) }

    elevator.finish()

    println(elevator.count)
  }
}